### DISCUSSION TITLE
Agent Session Based Workflow & Management Concept

### DISCUSSION THREAD

**Discussion by ArcTens** [microsoft/autogen#1779] [Ideas]
_Regarding Autogen & Autogen Studio_
tldr; add functionality to autogen to execute workflows across individual agent sessions to leverage memory and context to improve output quality.    

**Problem:**  LLM / API Text gen quality substantial increases over the course of individual sessions due to user feedback and context.  Some models are heavy on pleasantries/ monologues / unnecessary disclaimers and wordy explanations which rapidly go away over the course of an individual session, likely due to substantial context/memory taking over that space (maybe that's why some top tier pre-prompts are super long).  

**Autogen workflows can struggle to achieve that quality due to limitations of memory/context across workflows.**  

**Solution Concept:** Agent Sessions would be a new feature for autogen which would enable the creation and assignment of individual chat sessions to a workflow.  A workflow would essentially be executing across individual agent sessions in real time. 

![ui_concept_AGStudio-02](https://github.com/microsoft/autogen/assets/141588235/ec6967b7-4f2f-43d0-9f4f-909354342379)
![ui_concept_AGStudio-03](https://github.com/microsoft/autogen/assets/141588235/47f53d86-e209-45c1-9060-e61a18960233)
_(i do not recommend these models or this workflow its all improvised to convey the concept of the same agent being optimized for tasks via sessions)_

When configuring a workflow, the user would assign an agent as well as an individual agent session to the workflow, interactions with that agent happen within that session instance and the context/memory of that individual session is included along with the workflow context.  

After a workflow is completed, a user can navigate to any agent's individual session instance and has the options to tweak parameters and even message the agent in its own chat's log context (outside of a workflow).   This is the view my mockup concept was portraying. 

![ui_concept_AGStudio-01](https://github.com/microsoft/autogen/assets/141588235/938b5ef9-03d4-49fc-b5b1-195b76400374)

Options in this screen include the ability to edit, regenerate, or delete responses for the sake of improved context and testing.  Edits to the session do not roll up into the workflows own session/log.  Also the ability to assign/change Model at the session level is here as well.    

This can also be a way to **provide feedback to an agent**.  LLM memory context can degrade over time, so an option added to the mockup of additional pre-prompt insert can be a way to capture important information used for pre-prompting/memory.  Also a slider is available to adjust how much of the individual sessions memory/context is included when executing a workflow.

In the playground in Autogen studio an option also would be available to create a new individual 1 on 1 session with an agent.  this allows for testing of the agent's configuration as well as an ability to **temper an agent before it's assigned a workflow**.  

The end result would be providing the end user powerful tools to experiment with and improve their desired result with any individual agent, gaining substantially more milage out of models that other wise would need more tuning.  

The mockup below is only to demonstrate some functionality from a UI standpoint, workflow setting pictured on the left obviously wouldn't be located there but is demonstrating a dropdown to select a session when creating a workflow.  

**Interaction with an agent at session level can help with:**
- Carrying relevant context into future workflows
- Priming/Testing Agents before Workflow execution
- Expediting prompt optimization
- Troubleshooting undesired results 
- Regenerating a specific step of a workflow without having to re-run the entire workflow
- Model Testing/Swapping in an existing context

Just some ideas.

**Comment by ekzhu** [comment]
@ArcTens Have you shared this idea on Discord server? @victordibia do you have any comment on this idea?

**Reply by ArcTens** [reply]
I can do this

**Comment by ArcTens** [comment]
I may have done a poor job explaining here, but the concept can easily be proven in a wide variety of use cases simply by using a session based platform, such as chatgpt simply by going through an agent workflow, manually delegating agents as individual chat sessions.  you will quickly notice sessions with history in any given topic\task perform substantially better.   context is so powerful in correlation to output quality, i believe to be a critical missing piece for autogen

**Comment by ca9** [comment]
This is not an idea. This is a must. Not everything happens at runtime. Storing and recovering sessions is key for long term value
